build-scripts/npm-client.js

```diff
@@ -0,0 +1,183 @@
+ *   2. Publishes still succeed in the case that they were already published.
+ *       Allows Travis to publish on every commit.
+ */
+
+const {spawn} = require('child_process');
+const fs = require('fs');
+const path = require('path');
+
+// This script should catch and handle all rejected promises.
+// If it ever fails to do so, report that and exit immediately.
+process.on('unhandledRejection', error => {
+  console.error(error);
+  process.exit(1);
+});
+
+/**
+ * Simple function to log out publication process to a file. Standard out doesn't work because lerna
+ * swallows it.
+ *
+ * @param {string} message
+ */
+function logToFile(message) {
+  fs.writeFileSync(path.resolve(__dirname, '..', 'publish-log.txt'), `${message}\n`, {
+    flag: 'a' // Append to file
+  });
+}
+
+/**
+ * Execute a shell command as a promise which resolves to an Array of the form
+ *     [standardOut: string, standardError: string, exitCode: number]
+ *
+ * @param {string} cmd
+ * @param {!Array<strings>} args
+ * @return {!Promise<!{stdout: string, stderr: string, exitCode: number}>}
+ */
+function runCommand(cmd, args) {
+  return new Promise((resolve, reject) => {
+    let stdout = '';
+    let stderr = '';
+
+    const externalProcess = spawn(cmd, args, {
+      stdio: 'pipe'
+    });
+    externalProcess.on('error', err => {
+      reject({stdout, stderr, exitCode: -1});
+    });
+    externalProcess.on('close', exitCode => {
+      if (exitCode != 0) {
+        reject({stdout, stderr, exitCode});
+      }
+      resolve({stdout, stderr, exitCode});
+    });
+    externalProcess.stdout.on('data', data => {
+      stdout += data.toString();
+    });
+    externalProcess.stderr.on('data', data => {
+      stderr += data.toString();
+    });
+  });
+}
+
+/**
+ * Use the `npm view` command to ensure that a package@version combination
+ * has already been published to the registry.
+ *
+ * @param {string} packageWithVersion as specified in a package.json file "name@version"
+ * @return {!Promise<undefined>}
+ */
+function checkThatVersionExists(packageWithVersion) {
+  return runCommand('npm', ['view', packageWithVersion])
+      .catch(results => {
+        logToFile(results.stderr);
+        return Promise.reject(new Error('version did not exist in registry'));
+      });
+}
+
+/**
+ * Publish the package using `npm publish`.
+ *
+ * Ensure that any dependencies on other packages in this repo have already been published.
+ * Allows for the fact that the Graal images are published from separate Travis builds.
+ * Whichever run finishes last, publishes the main package.
+ *
+ * If the publish fails with an error that indicates the package has already been published,
+ * resolve the promise.
+ *
+ * @param {!Object<string, string|!Array|boolean>} packageInfo from the package.json file
+ * @return {!Promise<undefined>}
+ */
+function npmPublish(packageInfo) {
+  // Check both the "dependencies" and "optionalDependencies" keys
+  const depsInfo = [];
```

Comment:
What is supposed to go into depsInfo?
Looking below I'm guessing the elements are Promise of something.